### ISSUE TITLE
Find dialog keyboard accelerators can be broken by localization

### ISSUE THREAD

**Issue by Uhf7** [notepad-plus-plus/notepad-plus-plus#8696]
### Description of the Issue
If I use the `english_customizable.xml` localization, then two buttons in the Find/Replace dialog have **Alt-R** as accelerator:
- _Replace_ in the _Replace_ page.
- _Replace in Files_ in the _Find in Files page_.

It happens, that I have the _Replace_ page open, and with pressing **Alt-R** in the _Replace_ page I start the _Replace in Files_ function from the _Find in Files page_. Which has a safety question, which made me happy here.

### Steps to Reproduce the Issue

1. Start Notepad++ and set localization to _English (customizable)_.
2. Open the Find/Replace dialog.
3. Verify, that the _Replace_ button in the _Replace_ page has an underlined "R" (sometimes it becomes visible only after pressing the Alt key).
4. Verify, that the _Replace in Files_ button in the _Find in Files_ page has an underlined "R".
5. Go back to the _Replace_ page. 
6. Fill in something into the _Find what_ and _Replace with_ fields, like _A_ and _A_.
7. Press **Alt-R**. The function works like expected.
8. Use the **Tab** key to move the input focus to the _Replace_ button (in the _Replace_ page).
9. Press **Alt-R**. again.

### Expected Behavior
A normal _Replace_ function as happened in step 7.

### Actual Behavior
The _Replace in Files_ function from the _Find in Files_ page starts.

### Debug Information
Notepad++ v7.8.9   (32-bit)
Build time : Jul 15 2020 - 20:26:50
Path : C:\Tools\Npp\notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Pro (64-bit) 
OS Version : 1607
OS Build : 14393.576
Current ANSI codepage : 1252
Plugins : none

**Comment by sasumner** [comment]
>Which has a safety question, which made me happy here.

👍

**Comment by sasumner** [comment]
So the duplication of accelerators IS a problem, currently, but perhaps the bigger problem is that a function on a tab that isn't currently shown shouldn't be possible to be invoked.

**Comment by sasumner** [comment]
Related: #7638

**Comment by Uhf7** [comment]
> a tab that isn't currently shown shouldn't be possible to be invoked.

This would avoid at least the completely unexpected action and make it safer. And then, I think, that the buttons of inactive pages should be disabled automatically. But this is only in my mind as possible solution. Right now, I do nothing about it.

And it has to be tested before, whether Windows is smart enough to pass through the accelerator of the enabled button, if multiple buttons have the same accelerator and only one of those buttons is enabled.

**Comment by sasumner** [comment]
I think if user assigns same accelerator (on a single tab), then that is a user error (and I don't really care what Windows does with it or what happens in N++).  But that's just one possible thing that this issue raises.

**Comment by Uhf7** [comment]
I meant, that the menu elements in the Find/Replace dialog are already made visible/invisible depending of which page is active with the `ShowWindow` function. And it could be, that visibility is not important for accelerator keys. So, additionally to `ShowWindow`, `EnableWindow` should be called. This should silence the accelerators of inactive pages, in theory.

The same accelerator in one tab is indeed a translation error, which I don't really care about too.

**Comment by sasumner** [comment]
>This should silence the accelerators of inactive pages, in theory.

Yes, I suppose so.
Do you want to take on this work? (don't feel like you have to, just because you opened the issue)
If not, I will give it a go.

**Comment by Uhf7** [comment]
I would be glad, if you would do it, I stumbled into this only by accident.

**Comment by donho** [comment]
@Uhf7 
> Actual Behavior
The Replace in Files function from the Find in Files page starts.

I cannot reproduce it by following your instructions (even I don't get why english_customizable.xml makes change, I did it).
Here is my debug info, maybe you can tell me what I'm missing:
```
Notepad++ v7.8.9   (32-bit)
Build time : Aug 23 2020 - 12:38:21
Path : C:\@+)$°*#!?*\notepad-plus-plus\PowerEditor\visual.net\..\bin\Notepad++.exe
Admin mode : OFF
Local Conf mode : ON
OS Name : Windows 10 Enterprise (64-bit) 
OS Version : 1903
OS Build : 18362.1016
Current ANSI codepage : 1252
Plugins : DSpellCheck.dll mimeTools.dll NppConverter.dll NppExport.dll 

```

**Comment by Uhf7** [comment]
>  I don't get why english_customizable.xml makes change

If "English" is used, the button texts and accelerators come, at least after restart, directly from the menu defined in `FindReplaceDlg.rc`:

	
	PUSHBUTTON "&Replace",IDREPLACE,268,38,91,14
	PUSHBUTTON  "Replace in Files",IDD_FINDINFILES_REPLACEINFILES,268,38,91,14

Only the _Replace_ function has the accelerator &R here.

If "English (customizable)" is used, the button texts and accelerators come from the `English_customizable.xml`:

	<Item id="1608" name="&amp;Replace"/>
	<Item id="1660" name="&amp;Replace in Files"/>

Here both, the _Replace_ function and the _Replace in Files_ function have the accelerator &R.

The problem is, that the dialog box does respond to accelerator keys, even if the dialog box elements providing this accelerator keys are invisible.

----

I just reproduced it again, using the `npp.7.8.9.bin.7z` package, unpacked in a new directory.

Most important steps to reproduce it are steps 3, 4 and 8. Steps 3 and 4 verify, that the same accelerator key is assigned to both functions. Step 8 is necessary, because the dialog box reacts differently depending on where the current input focus is. This seems to be a Windows behavior.

**Comment by donho** [comment]
OK, thank you for the explanation. I didn't run the 8th step. Now I can reproduce it.
We should remove also `&amp;` from `<Item id="1660" name="&amp;Replace in Files"/>` in this case.

**Comment by Uhf7** [comment]
> We should remove also `&amp;`  from `<Item id="1660" name="&amp;Replace in Files"/>` in this case.

Not necessarily. This would make the xml equal to the resource file, for this reason, yes. But we could also include the & in the resource file to make the resource file equal to the xml.

I don't think we can, or should, prevent using the same accelerator key for different functions on different pages in the Find/Replace dialog. What we must prevent is, that the user is on one page in this dialog, and pressing an accelerator key which is present on another page too, and then the function from this other page is invoked.

**Comment by sasumner** [comment]
>I don't think we can, or should, prevent using the same accelerator key for different functions on different pages in the Find/Replace dialog.

This is successfully resolved with PR that accompanies this issue.

>What we must prevent is, that the user is on one page in this dialog, and pressing an accelerator key which is present on another page too, and then the function from this other page is invoked.

This is successfully resolved with PR that accompanies this issue.

**Comment by sasumner** [comment]
@donho 

If desired, I can reevaluate all of the accelerator key assignments for the four tabs of the _Find_ window, and reassign them in better ways, at least for `english_customizable.xml`.

However, I was thinking it was better to leave them the way they are, for users that don't do their own hacking, and are used to the accelerators the way they currently exist.

**Comment by donho** [comment]
> Not necessarily. This would make the xml equal to the resource file, for this reason, yes. But we could also include the & in the resource file to make the resource file equal to the xml.

The `english.xml` & `english_customizable.xml` sould be matched with the current situation anyway.

> I don't think we can, or should, prevent using the same accelerator key for different functions on different pages in the Find/Replace dialog. What we must prevent is, that the user is on one page in this dialog, and pressing an accelerator key which is present on another page too, and then the function from this other page is invoked.

Agree with this point.

**Comment by donho** [comment]
@sasumner 

> If desired, I can reevaluate all of the accelerator key assignments for the four tabs of the Find window, and reassign them in better ways, at least for english_customizable.xml.

No, it's not necessary. But as I said in the previous comment, the english.xml & english_customizable.xml sould be matched with the current situation.

> However, I was thinking it was better to leave them the way they are, for users that don't do their own hacking, and are used to the accelerators the way they currently exist.

Yes, you're right. Just leave it as it is.

**Comment by sasumner** [comment]
>The english.xml & english_customizable.xml should be matched with the current situation anyway.

I've never fully understood why there are TWO files for English.
There isn't two files for any other language.

Some guesses:

My thinking was that english.xml should be the "master" and should have the same strings as the resource file and the default UI strings used in code files.  At least this is how I think about it and do my PRs, which have always been accepted this way.

At release time, english_customizable.xml should be made as a copy from english.xml (and that copy should be installed only for new users that don't already have an english_customizable.xml file).

english.xml should not be changed by the user on their system, but should be copied over when user installs an upgrade.  english_customizable.xml is the one that users should change if they want, and it should NOT be updated during an upgrade, to avoid destroying previous user changes.  Users that customize need to look at diffs between a new english.xml and their changed english_customizable.xml (when they upgrade versions) in order to copy over new strings as software development adds them.

A reason for user not to change english.xml is that setting it for the localization is their only way to get their UI back to English, say for technical support reasons (bug reports should always be discussed in the English version).

Am I correct about any of this?

**Comment by donho** [comment]
@sasumner 
> I've never fully understood why there are TWO files for English.

english.xml is master (oups, sorry for the #BLM movement) language and should not be changed for the sake of switching between 2 non-English localizations. For example, if an user switches his/her localization from French to Arabic, what it happens is French->English->Arabic (of course if english.xml is present). Then reason is in Arabic translation some items could not be translated, so these items which are not applied to Arabic will be in English but not in French.

English speakers can, of course, always use english.xml as their customized file. But usually they don't install localization files, so english.xml is absent for them. `english_customizable.xml`is sort of reminder/indicator for English speakers who need change some names of items in Notepad++.

**Comment by sasumner** [comment]
Changes from this issue released in Notepad++ version 7.9

**Comment by sasumner** [comment]
So after the release of 7.9 I set it up as my daily-use version, and I noticed that while the situation that this issue is concerned with is greatly improved, there are STILL problems when two checkboxes not on the same tab page of the Find dialog are given the same Alt+key accelerator.  As far as I can tell so far, it only affects checkboxes, not other control types.

Here's what happens:  The checkbox control on the current page is given focus, but its functionality can't be activated, that is, the checkbox can't be checked/unchecked.  Due to this (partial support), it seems like some Windows oddity that perhaps Notepad++ code won't be able to workaround?  (I'm on Win10).

If anyone has great ideas on solving this, hopefully once-and-for-all, I'm listening.
